crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -139,6 +159,58 @@ where
             is_valid
         })
     }
+
+    /// Watch block reorgs and update filters accordingly
+    async fn watch_reorgs<Events>(&self, events: Events)
+    where
+        Events: CanonStateSubscriptions,
+    {
+        let mut canon_state_notifications = events.subscribe_to_canonical_state();
+        while let Ok(notification) = canon_state_notifications.recv().await {
+            if let CanonStateNotification::Reorg { old, .. } = notification {
+                self.update_reorg(old.blocks()).await;
+            }
+        }
+    }
+
+    /// update a reorg block for all active filters
+    async fn update_reorg(&self, old_blocks: &BTreeMap<BlockNumber, SealedBlockWithSenders>) {
+        let mut reorg_blocks = HashMap::<B256, BlockNumber>::with_capacity(old_blocks.len());
+        for block in old_blocks.values() {
+            reorg_blocks.insert(block.header.hash(), block.header.header().number);
+        }
+
+        for active_filter in self.active_filters().inner.lock().await.values_mut() {
+            if let FilterKind::Log(filter) = &active_filter.kind {
+                match filter.block_option {
+                    FilterBlockOption::AtBlockHash(block_hash) => {
+                        if let Some(block_number) = reorg_blocks.get(&block_hash) {
+                            active_filter.reorg_blocks.write().insert(block_hash, *block_number);
```

Comment:
I believe this write lock is okay because we already acquired the write lock on the filter